[New] `nvm use`/`nvm install`: add `--save` option

maartin0 opened the pull request:
In this PR I've:

 - added a `--save` option as per #2849 
 - implemented it for both `nvm use` and `nvm install` (which actually just carries over the option)
 - manually tested it in `bash`, `sh` and `zsh`
 - written one test, however I can't get the slow tests to run on my system so hopefully it runs okay :crossed_fingers: (if someone would be able to check that, that'd be great)

![image](https://user-images.githubusercontent.com/16228305/187242534-306fe814-c428-4181-a185-69a1355f8609.png)
![image](https://user-images.githubusercontent.com/16228305/187242305-0590d90c-5023-47b4-95a7-7e435beb2a2f.png)

Fixes #2849

ljharb commented on nvm.sh:
```diff
@@ -3155,6 +3157,10 @@ nvm() {
             shift
           ;;
+          --save)
+            NVM_WRITE_TO_NVMRC=true
```
this variable needs to be declared and initialized

ljharb commented on nvm.sh:
```diff
@@ -3573,11 +3580,16 @@ nvm() {
       fi
-
+      NVMRC_SAVE_TEXT=''
+      if [ "$NVM_WRITE_TO_NVMRC" = true ] && [ ! "_${VERSION}" = "_∞" ] && [ ! "${VERSION}" = 'N/A' ]; then
```
```suggestion
      if [ "${NVM_WRITE_TO_NVMRC}" = true ] && [ "${VERSION}" != '∞'  ] && [ "${VERSION}" != 'N/A' ]; then
```

ljharb reviewed (commented):
Let's add a bunch more tests - here's some example input:
 - `nvm use --save --lts`
 - `nvm use --save lts/argon`
 - `nvm use --save lts/*`
 - `nvm use --save node`
 - `nvm use --save iojs`
 - `nvm use --save foo` (where `foo` is a custom user alias)
 - `nvm use --save` (where the version is read from a `.nvmrc` in a higher directory)
 
 Let's also add some error tests for the cases where the `.nvmrc` file fails to be written (no permissions in the current directory, or the disk is full, for example)

ljharb commented on nvm.sh:
```diff
@@ -3056,6 +3058,7 @@ nvm() {
       local ALIAS
       local NVM_UPGRADE_NPM
       NVM_UPGRADE_NPM=0
+      NVM_WRITE_TO_NVMRC=0
```
```suggestion
      local NVM_WRITE_TO_NVMRC
      NVM_WRITE_TO_NVMRC=0
```
this is needed for dash

ljharb commented on nvm.sh:
```diff
@@ -3530,6 +3537,7 @@ nvm() {
       local NVM_LTS
       local IS_VERSION_FROM_NVMRC
       IS_VERSION_FROM_NVMRC=0
+      NVM_WRITE_TO_NVMRC="${NVM_WRITE_TO_NVMRC:=0}"
```
```suggestion
      local NVM_WRITE_TO_NVMRC
      NVM_WRITE_TO_NVMRC="${NVM_WRITE_TO_NVMRC:=0}"
```

ljharb commented on nvm.sh:
```diff
@@ -3573,11 +3582,15 @@ nvm() {
         return 127
       fi
-
+      NVMRC_SAVE_TEXT=''
```
```suggestion
      local NVMRC_SAVE_TEXT
      NVMRC_SAVE_TEXT=''
```

maartin0 commented:
Thanks for all the advice, it's been really helpful!
With the tests, would you suggest making seperate files for each one or loosely grouping them?

ljharb commented:
If you can make them be "fast" tests, then one file is totally fine - mainly it'll depend on how much setup is required. Generally each test file should only have one batch of setup at the beginning, and one batch of cleanup at the end.

maartin0 commented:
I've more or less implemented all of the tests, I ended up making them 'slow' because the only way of reliably testing all of the version aliases is by running `install` rather than `use`

I'm a little bit stuck on the last error test- how would you suggest checking if the command *didn't* fail? Just checking the exit code doesn't work because it seems like the exit code is invalid, for example
```bash
[ "$(nvm use --save --lts)" -eq 0 ]
```
gives me a weird error from `test` about it not being a number,
```bash
if nvm install --save --lts; then
```
doesn't trigger, even if an error message is thrown (I've tried using both `nvm_err` and `return 3`)

Other than that, do those other tests look okay?

maartin0 commented on nvm.sh:
```diff
@@ -3573,11 +3584,16 @@ nvm() {
+      local NVMRC_SAVE_TEXT
+      NVMRC_SAVE_TEXT=''
+      if [ "${NVM_WRITE_TO_NVMRC}" -eq 1 ] && [ "${VERSION}" != '∞' ] && [ "${VERSION}" != 'N/A' ]; then
+        echo "$VERSION" | tee "${PWD}"/.nvmrc > /dev/null || nvm_err "WARNING: Unable write version to .nvmrc"
```
Should this line `exit 3` on error or is a warning sufficient?

ljharb replied:
Yes, i think it should.

ljharb commented on nvm.sh:
```diff
@@ -3587,7 +3587,9 @@ nvm() {
       if [ "${NVM_WRITE_TO_NVMRC}" -eq 1 ] && [ "${VERSION}" != '∞' ] && [ "${VERSION}" != 'N/A' ]; then
-        echo "$VERSION" | tee "${PWD}"/.nvmrc > /dev/null || nvm_err "Unable write version to .nvmrc"
+        echo "$VERSION" | tee "${PWD}"/.nvmrc > /dev/null || {
+          nvm_err "WARNING: Unable write version to .nvmrc"
```
should this exit nonzero?

ljharb commented on nvm.sh:
```diff
@@ -3573,11 +3584,16 @@ nvm() {
         if nvm_has_system_node && nvm deactivate "${NVM_SILENT_ARG-}" >/dev/null 2>&1; then
           if [ "${NVM_SILENT:-0}" -ne 1 ]; then
-            nvm_echo "Now using system version of node: $(node -v 2>/dev/null)$(nvm_print_npm_version)"
+            nvm_echo "Now using system version of node: $(node -v 2>/dev/null)$(nvm_print_npm_version)$NVMRC_SAVE_TEXT"
```
```suggestion
            nvm_echo "Now using system version of node: $(node -v 2>/dev/null)$(nvm_print_npm_version)${NVMRC_SAVE_TEXT}"
```

ljharb commented on nvm.sh:
```diff
@@ -3633,7 +3649,7 @@ nvm() {
           NVM_USE_OUTPUT="Now using io.js $(nvm_strip_iojs_prefix "${VERSION}")$(nvm_print_npm_version)"
         else
-          NVM_USE_OUTPUT="Now using node ${VERSION}$(nvm_print_npm_version)"
+          NVM_USE_OUTPUT="Now using node ${VERSION}$(nvm_print_npm_version)$NVMRC_SAVE_TEXT"
```
```suggestion
          NVM_USE_OUTPUT="Now using node ${VERSION}$(nvm_print_npm_version)${NVMRC_SAVE_TEXT}"
```

ljharb commented on test/slow/nvm use/Running 'nvm use --save --lts' should work as expected:
```diff
@@ -0,0 +1,69 @@
+clean
+##
+
+nvm alias 'test' node
+nvm use 'test' --save || die 'nvm use test --save failed'
```
```suggestion
nvm alias test node || die '`nvm alias test node` failed'
nvm use test --save || die '`nvm use test --save` failed'
```

ljharb commented on nvm.sh:
```diff
@@ -3573,16 +3584,25 @@ nvm() {
+      NVMRC_SAVE_TEXT=''
+      if [ "${NVM_WRITE_TO_NVMRC}" -eq 1 ] && [ "${VERSION}" != '∞' ] && [ "${VERSION}" != 'N/A' ]; then
+        echo "$VERSION" | tee "${PWD}"/.nvmrc > /dev/null || {
+          nvm_err "Unable to write version to .nvmrc, aborted."
```
```suggestion
          nvm_err "Unable to write version \`${VERSION}\` to .nvmrc, aborted."
```

ljharb commented on nvm.sh:
```diff
@@ -3573,16 +3584,25 @@ nvm() {
+      local NVMRC_SAVE_TEXT
+      NVMRC_SAVE_TEXT=''
+      if [ "${NVM_WRITE_TO_NVMRC}" -eq 1 ] && [ "${VERSION}" != '∞' ] && [ "${VERSION}" != 'N/A' ]; then
+        echo "$VERSION" | tee "${PWD}"/.nvmrc > /dev/null || {
```
```suggestion
        echo "${VERSION}" | tee "${PWD}"/.nvmrc > /dev/null || {
```

ljharb commented on nvm.sh:
```diff
@@ -3573,16 +3584,25 @@ nvm() {
+          nvm_err "Unable to write version to .nvmrc, aborted."
+          exit 3
+        }
+        NVMRC_SAVE_TEXT=" (saved to .nvmrc)"
```
```suggestion
        NVMRC_SAVE_TEXT=' (saved to .nvmrc)'
```

ljharb commented on nvm.sh:
```diff
@@ -3573,16 +3584,25 @@ nvm() {
         if nvm_has_system_node && nvm deactivate "${NVM_SILENT_ARG-}" >/dev/null 2>&1; then
           if [ "${NVM_SILENT:-0}" -ne 1 ]; then
-            nvm_echo "Now using system version of node: $(node -v 2>/dev/null)$(nvm_print_npm_version)"
+            nvm_echo "Now using system version of node:
+            $(node -v 2>/dev/null)$(nvm_print_npm_version)${NVMRC_SAVE_TEXT}"
```
```suggestion
            nvm_echo "Now using system version of node: $(node -v 2>/dev/null)$(nvm_print_npm_version)${NVMRC_SAVE_TEXT}"
```
i don't think we want the node version on a separate line, that encourages people to parse it out of the message :-)

ljharb commented on test/slow/nvm use/Running 'nvm use --save --lts' should work as expected:
```diff
@@ -0,0 +1,82 @@
+set -ex
+
+clean () {
+  if [ "$(basename "$(pwd)")" = "tmp" ]; then
+    cd ..
+  fi
+  rm -rf .nvmrc tmp;
```
maybe we should store the directory path in a var, create it immediately, and fail if it doesn't work - and then use the var in `clean`? that way we also wouldn't have to rely on someone not running the tests in a folder also called "tmp"

If you make an entire folder for `--save`, then you could do this in `setup` and `teardown` files

ljharb commented on test/slow/nvm use/Running 'nvm use --save --lts' should work as expected:
```diff
@@ -0,0 +1,82 @@
+test_version () {
+  clean
+  VERSION=${1-}
+  nvm install --save "$VERSION" || die "\`nvm install --save $VERSION\` failed"
```
```suggestion
  nvm install --save "${VERSION}" || die "\`nvm install --save ${VERSION}\` failed"
```

ljharb commented on test/slow/nvm use/Running 'nvm use --save --lts' should work as expected:
```diff
@@ -0,0 +1,82 @@
+  VERSION=${1-}
+  nvm install --save "$VERSION" || die "\`nvm install --save $VERSION\` failed"
+  OUTPUT="$(cat .nvmrc)"
+  EXPECTED_OUTPUT=$(nvm_ls_current)
```
```suggestion
  EXPECTED_OUTPUT="$(nvm_ls_current)"
```
just in case it's an error message of some kind

ljharb commented on test/slow/nvm use/Running 'nvm use --save --lts' should work as expected:
```diff
@@ -0,0 +1,82 @@
+  OUTPUT="$(cat .nvmrc)"
+  EXPECTED_OUTPUT=$(nvm_ls_current)
+
+  [ "_$OUTPUT" = "_$EXPECTED_OUTPUT" ] \
+    || die "\`nvm use use --save $VERSION\`+ \`cat .nvmrc\` did not output '$EXPECTED_OUTPUT'; got '$OUTPUT'"
```
```suggestion
  [ "_${OUTPUT}" = "_${EXPECTED_OUTPUT}" ] \
    || die "\`nvm use use --save $VERSION\`+ \`cat .nvmrc\` did not output '${EXPECTED_OUTPUT}'; got '${OUTPUT}'"
```

ljharb commented on test/slow/nvm use/Running 'nvm use --save --lts' should work as expected:
```diff
@@ -0,0 +1,82 @@
+clean
+##
+
+nvm alias test node
+nvm use test --save || die "\`nvm use test --save\` failed"
+OUTPUT="$(cat .nvmrc)"
+EXPECTED_OUTPUT=$(nvm_ls_current)
+[ "_$OUTPUT" = "_$EXPECTED_OUTPUT" ] \
+  || die "\`nvm use use --save $VERSION\` + \`cat .nvmrc\` did not output '$EXPECTED_OUTPUT'; got '$OUTPUT'"
+nvm unalias test
```
might be nice to put each of these tests in their own file - ie, so `clean` is only needed once per file, at the end

maartin0 commented:
> Also, could the tests all be moved to `fast`? they'd be able to use mocks for `nvm ls-remote` calls, and as long as the dynamically-looked-up expected versions were fake-installed already, then they wouldn't need to actually install anything.

I've moved all of the test to 'fast' and split them into different files, however I'm not sure how to do fake installs (have you got any existing functions to do that?)

maartin0 commented on test/fast/Running 'nvm install --save --lts' with incorrect file permissions fails nicely:
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+set -e
```
Most tests have `set -ex` which makes sense generally, but `nvm install` makes huge amounts of output with this enabled, do you think `set -e` is sufficient?

ljharb replied:
i think that's fine; the `x` is to make debugging easier when it _does_ fail.

ljharb commented:
> however I'm not sure how to do fake installs (have you got any existing functions to do that?)

i do! grep for `make_fake_node`

maartin0 commented:
> > however I'm not sure how to do fake installs (have you got any existing functions to do that?)
> 
> i do! grep for `make_fake_node`

Hmm, I tried that (from `common.js`) and it didn't seem to prevent downloading the latest version

maartin0 commented:
If you'd like I can prepend that to all of the `nvm install` statements anyway

ljharb commented:
ah right, the other piece is that you have to mock out `nvm_ls_remote` - see https://github.com/nvm-sh/nvm/blob/master/test/fast/Unit%20tests/nvm%20ls-remote#L59-L66 for an example.

maartin0 commented:
I'm not quite sure how to mock `nvm_ls_remote`, so I've added 3 todo notes as to where  it needs to be ran;
All of the tests run fine, just some of them need to download specific versions of node

ljharb commented:
I added the mocks; I think this might be good to go!

It'll have to wait until the broken travis tests on master are fixed; please keep the fork/branch undeleted and the PR open until then :-)

ljharb commented:
I'll get the tests fixed and get this landed shortly.

maartin0 commented:
Any news with getting this merged?

ljharb commented:
@maartin0 sadly no; with the LTS promotion of node 18, tests are failing on master again.

maartin0 commented:
Any ideas why the Travis check is still "Waiting for status to be reported"?

ljharb commented:
Yes; Travis is a bit broken right now. I’ll handle it when I’ve got tests passing on master again.

ljharb commented:
It seems like there are in fact real test failures here - @maartin0, can you check out the travis test logs?

ljharb commented:
@maartin0 you can ignore the WSL failures, but the travis failures seem real - you may need to remove the local `.nvmrc` after your `--save` tests, to leave a clean slate? (be sure to back up an existing one if present, and then restore it afterwards)

maartin0 commented:
> @maartin0 you can ignore the WSL failures, but the travis failures seem real - you may need to remove the local `.nvmrc` after your `--save` tests, to leave a clean slate? (be sure to back up an existing one if present, and then restore it afterwards)

Ok thanks, sorry for the *slightly late* reply - from what I could see, there were a couple of issues so I've:
* Moved the logic in the "nvm install" branch slightly earlier so it'll still run even if the specified version of node is already installed (which seemed to be the cause of some of the tests failing)
* Changed the output to use `nvm_echo "Wrote version number (${VERSION}) to .nvmrc"` rather than appending it to other outputs which seemed a bit untidy
* Replaced the test that runs `nvm install --save lts/argon` with `lts/iron` since it randomly failed on my system, I'm not completely sure why but I'd guess it's todo with the fact that it's a pretty old version of node, but if one works then the other should technically work
* As requested, made each test backup any existing `.nvmrc` to `.nvmrc.orig` then restores it afterwards

The three tests that are still failing are due to a weird error I think after calling `make_fake_node node`, this is what the output looks like running on my system (WSL Ubuntu) and you can see the first one that fails is when calling `nvm install --save node` but it works fine when calling it normally without the `make_fake_node` call. 
```
> ./"Running 'nvm install --save' works as expected"
+ nvm install --save --lts
Installing latest LTS version.
Wrote version number (v21.1.0) to .nvmrc
v21.1.0 is already installed.
Now using node v21.1.0 (npm v10.2.0)
+ nvm install --save lts/iron
Wrote version number (v21.1.0) to .nvmrc
v21.1.0 is already installed.
Now using node v21.1.0 (npm v10.2.0)
+ nvm install --save lts/*
Wrote version number (v21.1.0) to .nvmrc
v21.1.0 is already installed.
Now using node v21.1.0 (npm v10.2.0)
+ nvm install --save node
Wrote version number (v0.1.14) to .nvmrc
Detected that you have 8 CPU core(s)
Running with 7 threads to speed up the build
Local cache found: ${NVM_DIR}/.cache/src/node-v0.1.14/node-v0.1.14.tar.gz
Checksums match! Using existing downloaded archive ${NVM_DIR}/.cache/src/node-v0.1.14/node-v0.1.14.tar.gz
$>./configure --prefix=/home/martin/.nvm/v0.1.14 <
Traceback (most recent call last):
  File "/home/martin/.nvm/.cache/src/node-v0.1.14/files/tools/waf-light", line 147, in <module>
    import Scripting
  File "/home/martin/.nvm/.cache/src/node-v0.1.14/files/tools/wafadmin/Scripting.py", line 143
    except Utils.WafError, e:
           ^^^^^^^^^^^^^^^^^
SyntaxError: multiple exception types must be parenthesized
nvm: install v0.1.14 failed!
`nvm install --save node` failed
> \. ../../../nvm.sh
> nvm install --save node
Wrote version number (v21.6.2) to .nvmrc
v21.6.2 is already installed.
Now using node v21.6.2 (npm v10.2.4)
>
```
I'm not sure if this is a weird aliasing issue or something but if you can't see an easy way around it do you think I should get rid of/just ignore the `node` test for this unit and then use `lts/*` for the other two that are failing?

ljharb commented:
I intentionally use nonexistent node versions in the tests at some point; make_fake_node should be working fine already.

ljharb commented on nvm.sh:
```diff
@@ -3350,6 +3358,15 @@ nvm() {
         FLAVOR="$(nvm_node_prefix)"
       fi
 
+      # Write .nvmrc if flag is set
+      if [ $NVM_WRITE_TO_NVMRC -eq 1 ] && [ "${VERSION}" != '∞' ] && [ "${VERSION}" != 'N/A' ]; then
+        echo "${VERSION}" | tee "${PWD}"/.nvmrc > /dev/null || {
+          nvm_err "Warning: Unable to write version number (${VERSION}) to .nvmrc"
+          exit 3
+        }
+        nvm_echo "Wrote version number (${VERSION}) to .nvmrc"
+      fi
```
this should only happen if the version is already installed, or is successfully installed - not if installation fails.

ljharb commented on nvm.sh:
```diff
@@ -3650,6 +3671,19 @@ nvm() {
         return 127
       fi
 
+      # Write .nvmrc if flag is set
+      if [ $NVM_WRITE_TO_NVMRC -eq 1 ] && [ "${VERSION}" != '∞' ] && [ "${VERSION}" != 'N/A' ]; then
+        echo "${VERSION}" | tee "${PWD}"/.nvmrc > /dev/null || {
+          if [ "${NVM_SILENT:-0}" -ne 1 ]; then
+            nvm_err "Warning: Unable to write version number (${VERSION}) to .nvmrc"
+          fi
+          exit 3
+        }
+        if [ "${NVM_SILENT:-0}" -ne 1 ]; then
+          nvm_echo "Wrote version number (${VERSION}) to .nvmrc"
+        fi
+      fi
```
similarly, this shouldn't happen until the `use` is successful.

maartin0 commented:
I've refactored it a bit by moving the block under `nvm_write_nvmrc` and then only running it on `nvm use` (I removed the `local` declaration in the `nvm use` block so `NVM_WRITE_TO_NVMRC` is passed through). Can you figure out why the tests aren't working? They run fine as independent shells scripts or by running `urchin` directly in the `test/fast/Unit tests` directory, but they fail as soon as you try running it further up and I can't get any error messages out of it- is something in one of the `setup` or `setup_dir` scripts interfering with it?

ljharb commented on nvm.sh:
```diff
@@ -3132,6 +3151,7 @@ nvm() {
       local ALIAS
       local NVM_UPGRADE_NPM
       NVM_UPGRADE_NPM=0
+      local NVM_WRITE_TO_NVMRC
```
```suggestion
      local NVM_WRITE_TO_NVMRC
      NVM_WRITE_TO_NVMRC=0
```
to make sure it's set in every code path

maartin0 replied:
I took away that declaration so that the `NVM_WRITE_TO_NVMRC` from `nvm install` stays set when `nvm use` gets called, which I thought was better than appending `--save` to each of the `nvm use` calls within the `nvm install` block

maartin0 replied:
Oh never mind, that's not what you were referring to - can do although it shouldn't make a difference since I'm just using a default arg when checking (`[ "${NVM_WRITE_TO_NVMRC:-0}" -eq 1 ]`)

ljharb reviewed (commented):
This is looking pretty good assuming tests pass!

maartin0 commented:
@ljharb The issue with the tests was that I was using `set -e` when sourcing nvm.sh which I only just realised you haven't done in any of the other tests in `fast/Unit tests` - is that ok? It seems a bit dodgy

maartin0 commented:
Just pushed again with a 3 more things:

* Added `-w` alias for `--save` and a test to go with it
* Fixed a *bug* in `nvm_auto` where it tries to resolve the alias `default` which resolves to `N/A` which is a non-empty string but invalid version so when it calls `nvm use` or `nvm install` it'll `exit 3` if `set -e` is set -- this has got the number of failing `fast` tests from 7 down to 2
* Added `set -e` flags back to the tests I wrote since `nvm.sh` can be sourced successfully now

ljharb commented:
> @ljharb The issue with the tests was that I was using `set -e` when sourcing nvm.sh which I only just realised you haven't done in any of the other tests in `fast/Unit tests` - is that ok? It seems a bit dodgy

yeah, that's fine - sometimes i have to `set -e` *after* sourcing nvm.sh.

setting -e isn't even really a good practice anymore, so i'm not too worried about it.

maartin0 commented:
Just rebased again if you think this is ready to merge?

ljharb commented:
Unfortunately the Travis tests are still failing - https://app.travis-ci.com/github/nvm-sh/nvm/jobs/619295015

maartin0 commented:
@ljharb There was only one issue which was surprisingly simple to fix, but annoying to figure out - I just needed to add a couple of extra bits to the `cleanup` functions in all the tests, since they were leaving behind installed versions which the other tests didn't expect, which in turn changed the output format of `nvm ls-remote` to literally just have one blue rather than black line:
![nvm ls-remote test expected/actual output](https://github.com/nvm-sh/nvm/assets/16228305/2d560812-d2a7-43eb-9028-2ee1cfaa60c5)
And this is what I needed to change `cleanup` to (which I stole from [`nvm_install_no_progress_bar`](https://github.com/nvm-sh/nvm/blob/master/test/fast/Unit%20tests/nvm_install_no_progress_bar))
```sh
cleanup () {
  nvm cache clear
  nvm deactivate
  nvm unalias default
  rm -rf ${NVM_DIR}/v* .nvmrc
  if [ -f .nvmrc.orig ]; then mv .nvmrc.orig .nvmrc; fi
  unset -f nvm_ls_remote nvm_ls_remote_iojs
}
```

ljharb commented:
@maartin0 sorry for the delay here; can you please rebase this and resolve the conflicts?

maartin0 commented:
@ljharb All done and no worries, although Travis seems to have another issue?
> $ git clone --depth=50 https://github.com/nvm-sh/nvm.git nvm-sh/nvm
> $ git submodule update --init --recursive --quiet
> Warning: Permanently added the ECDSA host key for IP address '140.82.112.4' to the list of known hosts.
> git@github.com: Permission denied (publickey).
> fatal: Could not read from remote repository.

ljharb commented:
yep, known and don’t worry about that one :-)
